File: Python/125_Valid_Palindrome.py
"""
125. Valid Palindrome
Easy

Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Note: For the purpose of this problem, we define empty string as valid palindrome.

Example 1:

Input: "A man, a plan, a canal: Panama"
Output: true

Example 2:

Input: "race a car"
Output: false

"""

class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True
        
        string = ''.join(c.lower() for c in s if c.isalnum())
        return string == string[::-1]

    def isPalindrome2(self, s: str) -> bool:

        s = s.lower()
        s = ''.join(list(filter(lambda x: x.isalnum(), s)))
        return s == s[::-1]

File: String/125_Valid_Palindrome.cpp
/*************************************************************************
125. Valid Palindrome
Easy

Given a string, determine if it is a palindrome, considering only alphanumeric
 characters and ignoring cases.
For example,
”A man, a plan, a canal: Panama” is a palindrome.
”race a car” is not a palindrome.

Note: Have you consider that the string might be empty? 
This is a good question to ask during an interview.
For the purpose of this problem, we define empty string as valid palindrome.
*************************************************************************/

class Solution {
public:
  bool isPalindrome(string s) {

  	int left = 0, right = s.size()-1;
  	while(left < right) {
  		if(!isalnum(s[left])) left ++;
  		else if(!isalnum(s[right])) right --;
  		else if(tolower(s[left]) != tolower(s[right])) return false;
		else {
  			left ++; right --;
		}
  	}
  	return true;
  }
};